### ISSUE TITLE
Error analysis 

### ISSUE THREAD

**Issue by iLOSTA** [iLOSTA/The-Project#10]
Although the current model's accuracy is pretty good (99.7%), more work still needs to be done to identify the possible source of the remaining 0.03%. To do so, I made predictions for the entire dataset, and saved the incorrectly predicted samples. 

Out of the 3344 samples, 56 samples were incorrectly classified. Within the 56 errors, 34 samples had errors in the MODEL category and 16 samples had errors in the PANEL category. some samples had errors in more than one category. Details in the graph below.

![Capture](https://user-images.githubusercontent.com/50953458/61089247-64531f80-a432-11e9-8faa-26d9cf7e637c.PNG)

The next graph shows which models and panels and other categories were more problematic. As can be seen, car models A2, Auris and Kadjar were the highest error causes. in the panel category, Bonnet was incorrectly classified 8 times and Boot 4 times. 
![details](https://user-images.githubusercontent.com/50953458/61089739-06273c00-a434-11e9-82db-c8f79974b36f.png)

**Comment by iLOSTA** [comment]
UPDATE 1:

I did a little more investigation around the possible sources of errors and how to mitigate them. I used three different variations of the Tf-idf vectorizer: word (1,2) grams, char (2,4) grams and char_wb (2,4) grams. I noticed:
- The number of errors is about the same for all three methods (word: 62 errors, char : 71 errors, char_wb: 74 errors).
- However, the distribution of those errors is different. Looking at the top 3 error sources categories namely MODEL, MAKE and PANEL, I found out: 
-- in the case of word(1,2) there are 36 errors in MODEL, 12 in MAKE, 13 in PANEL
-- in the case of char(2,4) there are 44 errors in MODEL, 16 in MAKE, 13 in PANEL
-- but in char_wb(2,4), there are     45 error in MODEL, 8 in MAKE, 7 in PANEL
- Looking at the case of char_wb, although it produced 8 more errors compare to word(1,2), the errors are mostly coming from one source which is MODEL. 
![char_wb24_](https://user-images.githubusercontent.com/50953458/61163795-68ea0780-a508-11e9-833b-9c73f6aaa4f0.png)
More errors coming from the same source makes mitigating these errors easier. One solution is to generate more training samples focusing on CAR-MODELS Toyota Auris, Toyota Camry, Toyota Corolla, Nissan Juke as these models are the most problematic ones. Doing so should fix a big portion of the output error.

**Comment by iLOSTA** [comment]
UPDATE 2:

I created 42 new samples that focused on the most erroneous categories based on the previous error analysis step. These 42 samples were then turned into 153 samples by making the STT API return more than one result per audio file. The new samples were added together with the original dataset and the model was retrained on the new dataset. Adding the new samples, along with the following modifications to the structure of the NN model fixed some of the errors.
- Number of neurons in each layer was changed to 80-70-50 instead of 80-60-40 previously. The reason behind this change was that I thought the network was losing some of the features that it learned at each layer when moving to the next layer.
- The Optimizer that is used to compile the model was changed from _Adam_ to _Nadam_. Nadam is just an improved version of Adam. 
![EXTRA_](https://user-images.githubusercontent.com/50953458/61180576-42af8f00-a610-11e9-8a17-0fe885e17976.png)

As can be seen, the total number of errors was decreased from 74 to 47. Previously, 45 errors happened in the car-model section where now there are only 25. It is true that the number of errors in some other sections increased slightly, like in LOCATION_X and Y, but it is a trade off in order to improve the overall performance of the Neural Network Model.
